src/Common/src/System/Dynamic/Utils/CollectionExtensions.cs

```diff
@@ -57,24 +57,29 @@ public static int ListHashCode<T>(this IEnumerable<T> list)
         }
 
         [Pure]
-        public static bool ListEquals<T>(this ICollection<T> first, ICollection<T> second)
+        public static bool ListEquals<T>(this ReadOnlyCollection<T> first, ReadOnlyCollection<T> second)
         {
-            if (first.Count != second.Count)
+            if (first == second)
+            {
+                return true;
+            }
+
+            int count = first.Count;
+
+            if (count != second.Count)
             {
                 return false;
             }
+
             var cmp = EqualityComparer<T>.Default;
-            var f = first.GetEnumerator();
-            var s = second.GetEnumerator();
-            while (f.MoveNext())
+            for(int i = 0; i != count; ++i)
             {
-                s.MoveNext();
-
-                if (!cmp.Equals(f.Current, s.Current))
+                if (!cmp.Equals(first[i], second[i]))
```

Comment:
Presumably ReadOnlyCollection is only used in the codebase with IList implementations where indexing is faster than enumeration?